s3/s3_operations.cpp

```diff
@@ -0,0 +1,1682 @@
+
+                size_t retry_cnt = 0;
+                do {
+
+                    std::string&& hostname = s3GetHostname(_ctx.prop_map());
+                    bucketContext.hostName = hostname.c_str();
+                    data.pCtx = &bucketContext;
+
+                    S3_list_bucket(&bucketContext,                                       // S3BucketContext
+                            search_key.c_str(),                                          // prefix
+                            data.next_marker == "" ? nullptr : data.next_marker.c_str(), // marker
+                            "/",                                                         // delimiter
+                            1024,                                                        // max number returned
+                            nullptr,                                                     // S3RequestContext
+                            0,                                                           // timeout
+                            &list_bucket_handler,                                        // S3ListBucketHandler
+                            &cb_data                                                     // void* callback data
+                            );
+
+                    if (data.status != S3StatusOK) s3_sleep( retry_wait, 0 );
+
+                } while ( (data.status != S3StatusOK) && S3_status_is_retryable(data.status) && (++retry_cnt < retry_count_limit ) );
+
+                if (data.status != S3StatusOK) {
+                    std::stringstream msg;
+                    msg << "[resource_name=" << get_resource_name(_ctx.prop_map()) << "] ";
+                    msg << " - Error in S3 listing: \"" << search_key.c_str() << "\"";
+                    if (data.status >= 0) {
+                        msg << " - \"" << S3_get_status_name((S3Status)data.status) << "\"";
+                    }
+                    return ERROR(S3_FILE_STAT_ERR, msg.str());
+                }
+            }
+
+            *_dirent_ptr = nullptr;
+            if (data.returned_objects.size() > 0) {
+
+                std::string current_key = data.returned_objects.front();
+                data.returned_objects.pop_front();
+
+                // a trailing slash indicates a "directory", remove it
+                // for boost::filesystem::path can get dir name then add
+                // it back in
+                //bool add_trailing_slash = false;
+                //if('/' == current_key.back()) {
+                //    add_trailing_slash = true;
+                //    current_key.pop_back();
+                //}
+
+                *_dirent_ptr = ( rodsDirent_t* ) malloc( sizeof( rodsDirent_t ) );
+                boost::filesystem::path p(current_key.c_str());
+                current_key = p.filename().string();
+                //if (add_trailing_slash) {
+                //    current_key += "/";
+                //}
```

Comment:
let's make sure the comments are true and match the code behavior

Reply:
Need to test this out.  Will clean this up.

Reply:
After more inspection, this code is no longer needed as we now handle returned objects differently than returned collections.